rust/windows-client/src-tauri/src/client/gui.rs

```diff
@@ -403,6 +407,70 @@ impl Controller {
 
         Ok(())
     }
+
+    fn copy_resource(&self, id: &str) -> Result<()> {
+        let Some(session) = &self.session else {
+            bail!("app is signed out");
+        };
+        let resources = session.callback_handler.resources.load();
+        let id = ResourceId::from_str(id)?;
+        let Some(res) = resources.iter().find(|r| r.id() == id) else {
+            bail!("resource ID is not in the list");
+        };
+        let mut clipboard = arboard::Clipboard::new()?;
+        // TODO: Make this a method on `ResourceDescription`
+        match res {
+            ResourceDescription::Dns(x) => clipboard.set_text(&x.address)?,
+            ResourceDescription::Cidr(x) => clipboard.set_text(&x.address.to_string())?,
+        }
+        Ok(())
+    }
+
+    async fn handle_deep_link(&mut self, url: &url::Url) -> Result<()> {
+        let Some(auth) = client::deep_link::parse_auth_callback(url) else {
+            bail!("couldn't parse scheme request");
+        };
+
+        let token = auth.token.clone();
+        // spawn_blocking because keyring access is I/O
+        if let Err(e) = spawn_blocking(move || {
+            let entry = keyring_entry()?;
+            entry.set_password(token.expose_secret())?;
+            Ok::<_, anyhow::Error>(())
+        })
+        .await?
+        {
+            tracing::error!("couldn't save token to keyring: {e:#?}");
+        }
+
+        let auth_info = AuthInfo {
+            actor_name: auth.actor_name,
+            token: auth.token,
+        };
+        if let Err(e) = self.start_session(auth_info) {
+            bail!("couldn't start session: {e:#?}");
+        }
+        Ok(())
+    }
+
+    fn reload_resource_list(&mut self) -> Result<()> {
+        let Some(session) = &self.session else {
+            tracing::warn!("got notified to update resources but there is no session");
```

Comment:
I think this could be a `debug_assert`

Reply:
I was thinking because of the mpsc channel, I might be able to get a valid resources notification after the session has already ended